Editor: Move scenario components when resizing scenario's map

babaissarkar opened the pull request:
Resolves #10520

**TODO**
- [x] Move unit (undoable)
- [x] avoid unit suddenly showing idle animation after move
- [x] Move items
- [x] Move labels (undoable)
- [x] make item moving undoable (needs a new editor action)
- [x] move Time Areas

**NOTE**
Due to the way map coordinates work in hex grid, visual adjacency after resizing is not guaranteed.

Before:
<img width="208" height="165" alt="Screenshot from 2025-09-20 21-44-32" src="https://github.com/user-attachments/assets/6b1e40f5-dabf-4acf-a71f-8986f4ba6bc1" />
After resize (add 1 to width):
<img width="262" height="249" alt="Screenshot from 2025-09-20 21-44-40" src="https://github.com/user-attachments/assets/5b1d7929-ea10-4093-8966-bec3745af14c" />

gfgtdf commented:
> Due to the way map coordinates work in hex grid, visual adjacency after resizing is not guaranteed.

when resizing a map like that, does it only mess up the units or also the map itself? Like if there was for example a hill tile at the position of the dwarf would the hill also be not-adjacent to the water tile?

babaissarkar commented:
> also, i'd expect this implementation to not work with editor undoing since its not part of the editor action.

It does, I used unit replace action.

babaissarkar commented:
> > Due to the way map coordinates work in hex grid, visual adjacency after resizing is not guaranteed.
> 
> when resizing a map like that, does it only mess up the units or also the map itself? Like if there was for example a hill tile at the position of the dwarf would the hill also be not-adjacent to the water tile?

It may, in some cases, but the effect for terrain isn't very obvious. Notice the cobbles below.

<img width="229" height="195" alt="Screenshot from 2025-09-21 06-31-20" src="https://github.com/user-attachments/assets/99057620-d20c-4f32-9a84-a3e1c3d320f2" />
<img width="224" height="220" alt="Screenshot from 2025-09-21 06-31-12" src="https://github.com/user-attachments/assets/989c1b51-790a-4636-93c8-b330bbae1a9d" />

babaissarkar commented on src/editor/map/context_manager.cpp:
```diff
@@ -628,8 +634,48 @@ void context_manager::resize_map_dialog()
+		chain->append_action(std::make_unique<editor_action_item_move_all>(x_offset, y_offset));
+
+		map_labels& labels = get_map_context().get_labels();
+		get_map_context().map().for_each_loc([&](const map_location& loc) {
```
@Vultraz Need help trying to simplify this block. It works but adds way too many actions to the chain.

gfgtdf commented:
also: does it still work when units/labels etc are in a location that would be removed by the resizing operation?

babaissarkar commented:
> also: does it still work when units/labels etc are in a location that would be removed by the resizing operation?

I don't think that should cause an issue as everything is moved inwards usually AFAIK. If you have a specific test case I can try that.

babaissarkar commented:
> are [time_area]s also moved properly?

Implemented in https://github.com/wesnoth/wesnoth/pull/10568/commits/5ea86550bce413fd1b072fc74bde0a2ad4661d3a.

babaissarkar commented:
> > I don't think that should cause an issue as everything is moved inwards usually AFAIK. If you have a specific test case I can try that.
> 
> I'd think just use the resizing dialog to shrink the map (new size smaller than old size)

What would be the expected correct behavior?

rodolphecombe commented:
> Also undoing bringing these elements back correctly.
I would not expect that to work TBH, so, IMHO, it not working should not prevent the PR from being merged.
